From aef08bae45c743d49da45972cac898284203bc53 Mon Sep 17 00:00:00 2001
From: Vladimir Zapolskiy <vladimir_zapolskiy@mentor.com>
Date: Thu, 1 Oct 2015 16:36:16 +0300
Subject: [PATCH 665/668] media: imx6 capture: adv7180: check return value of
 v4l2_ctrl_handler_setup()

Don't forget to check return value of v4l2_ctrl_handler_setup() and
in case of an error free registered V4L2 controls. Note, hdlr->error
in this case may be not set accordingly.

Signed-off-by: Vladimir Zapolskiy <vladimir_zapolskiy@mentor.com>
---
 drivers/staging/media/imx6/capture/adv7180.c | 21 +++++++++++++--------
 1 file changed, 13 insertions(+), 8 deletions(-)

diff --git a/drivers/staging/media/imx6/capture/adv7180.c b/drivers/staging/media/imx6/capture/adv7180.c
index 13b53a1..a870ec6 100644
--- a/drivers/staging/media/imx6/capture/adv7180.c
+++ b/drivers/staging/media/imx6/capture/adv7180.c
@@ -1090,7 +1090,7 @@ static const struct v4l2_ctrl_config adv7180_custom_ctrl[] = {
 static int adv7180_init_controls(struct adv7180_dev *sensor)
 {
 	const struct v4l2_ctrl_config *c;
-	int i;
+	int i, err;
 
 	v4l2_ctrl_handler_init(&sensor->ctrl_hdl, ADV7180_NUM_CONTROLS);
 
@@ -1109,16 +1109,21 @@ static int adv7180_init_controls(struct adv7180_dev *sensor)
 
 	sensor->sd.ctrl_handler = &sensor->ctrl_hdl;
 	if (sensor->ctrl_hdl.error) {
-		int err = sensor->ctrl_hdl.error;
-
-		v4l2_ctrl_handler_free(&sensor->ctrl_hdl);
-
-		v4l2_err(&sensor->sd, "%s: error %d\n", __func__, err);
-		return err;
+		err = sensor->ctrl_hdl.error;
+		goto error;
 	}
-	v4l2_ctrl_handler_setup(&sensor->ctrl_hdl);
+
+	err = v4l2_ctrl_handler_setup(&sensor->ctrl_hdl);
+	if (err)
+		goto error;
 
 	return 0;
+
+ error:
+	v4l2_ctrl_handler_free(&sensor->ctrl_hdl);
+	v4l2_err(&sensor->sd, "%s: error %d\n", __func__, err);
+
+	return err;
 }
 
 static int adv7180_enum_framesizes(struct v4l2_subdev *sd,
-- 
2.4.10

